File: sdk/python/jamjet/cloud/outcomes.py
"""Record trace outcomes against the JamJet Cloud API."""

from __future__ import annotations

from typing import Any

import httpx

VALID_OUTCOMES = frozenset(
    {"success", "failure", "approved", "rejected", "resolved", "unresolved"}
)


def record_outcome(
    api_key: str,
    api_url: str,
    trace_id: str,
    outcome: str,
    score: float | None = None,
    metadata: dict[str, Any] | None = None,
) -> None:
    """POST an outcome record to ``/v1/outcomes``.

    Args:
        api_key:  Bearer token for the JamJet Cloud API.
        api_url:  Base URL, e.g. ``https://api.jamjet.dev``.
        trace_id: ID of the trace whose outcome is being recorded.
        outcome:  One of ``success``, ``failure``, ``approved``, ``rejected``,
                  ``resolved``, or ``unresolved``.
        score:    Optional float in [0, 1] representing a quality score.
        metadata: Optional free-form dict attached to the record.

    Raises:
        ValueError: When *outcome* is not one of the six valid values or
                    *score* is outside the [0, 1] range.
        httpx.HTTPStatusError: When the API returns a non-2xx response.
    """
    if outcome not in VALID_OUTCOMES:
        raise ValueError(
            f"Invalid outcome {outcome!r}. Must be one of: "
            + ", ".join(sorted(VALID_OUTCOMES))
            + "."
        )
    if score is not None:
        if not isinstance(score, (int, float)) or score < 0 or score > 1:
            raise ValueError(
                f"score must be a number between 0 and 1 (inclusive), got {score!r}."
            )

    payload: dict[str, Any] = {"trace_id": trace_id, "outcome": outcome}
    if score is not None:
        payload["score"] = float(score)
    if metadata is not None:
        payload["metadata"] = metadata

    resp = httpx.post(
        f"{api_url.rstrip('/')}/v1/outcomes",
        json=payload,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        timeout=10,
    )
    resp.raise_for_status()

File: sdk/typescript/packages/cloud/src/outcomes.ts
const VALID_OUTCOMES = new Set([
  'success',
  'failure',
  'approved',
  'rejected',
  'resolved',
  'unresolved',
] as const)

export type Outcome =
  | 'success'
  | 'failure'
  | 'approved'
  | 'rejected'
  | 'resolved'
  | 'unresolved'

export interface RecordOutcomeOptions {
  score?: number
  metadata?: Record<string, unknown>
}

function isValidOutcome(value: string): value is Outcome {
  return VALID_OUTCOMES.has(value as Outcome)
}

/**
 * Record the outcome of a trace run.
 *
 * POSTs to `POST /v1/outcomes` on the JamJet Cloud API.
 *
 * @param apiKey   Bearer token.
 * @param apiUrl   Base URL (e.g. `https://api.jamjet.dev`).
 * @param traceId  The trace whose outcome is being recorded.
 * @param outcome  One of the six outcome values.
 * @param opts     Optional `score` (0–1) and free-form `metadata`.
 */
export async function recordOutcome(
  apiKey: string,
  apiUrl: string,
  traceId: string,
  outcome: Outcome,
  opts: RecordOutcomeOptions = {},
): Promise<void> {
  if (!isValidOutcome(outcome)) {
    throw new TypeError(
      `Invalid outcome "${outcome}". Must be one of: ${[...VALID_OUTCOMES].join(', ')}.`,
    )
  }
  if (opts.score !== undefined) {
    if (typeof opts.score !== 'number' || opts.score < 0 || opts.score > 1) {
      throw new RangeError(
        `score must be a number between 0 and 1 (inclusive), got ${opts.score}.`,
      )
    }
  }

  const body: Record<string, unknown> = { trace_id: traceId, outcome }
  if (opts.score !== undefined) body.score = opts.score
  if (opts.metadata !== undefined) body.metadata = opts.metadata

  const res = await fetch(`${apiUrl}/v1/outcomes`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  })

  if (!res.ok) {
    throw new Error(`JamJet recordOutcome failed: ${res.status} ${res.statusText}`)
  }
}
